src/lib/mdns/TxtFields.cpp

```diff
@@ -0,0 +1,281 @@
+    for (size_t i = 0; i < val.size(); ++i)
+    {
+        char c = static_cast<char>(val.data()[i]);
+        if (c < '0' || c > '9')
+        {
+            return errval;
+        }
+        u32 = u32 * 10;
+        u32 += val.data()[i] - static_cast<uint8_t>('0');
+        if (u32 > std::numeric_limits<uint16_t>::max())
+        {
+            return errval;
+        }
+    }
+    return static_cast<uint16_t>(u32);
+}
+
+uint8_t MakeU8FromAsciiDecimal(const ByteSpan & val)
+{
+    uint16_t u16 = MakeU16FromAsciiDecimal(val);
+    if (u16 > std::numeric_limits<uint8_t>::max())
+    {
+        return 0x0;
+    }
+    return static_cast<uint8_t>(u16);
+}
+
+CHIP_ERROR MakeU8FromAsciiHex(const ByteSpan & ascii, uint8_t * val)
+{
+    if (ascii.size() != 2)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    uint8_t ret = 0;
+    for (size_t i = 0; i < ascii.size(); ++i)
+    {
+        ret    = ret << 4;
+        char c = static_cast<char>(ascii.data()[i]);
+        if (c >= '0' && c <= '9')
+        {
+            ret += ascii.data()[i] - static_cast<uint8_t>('0');
+        }
+        // Only uppercase is supported according to spec.
+        else if (c >= 'A' && c <= 'F')
+        {
+            ret += ascii.data()[i] - static_cast<uint8_t>('A') + 0xA;
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+    *val = ret;
+    return CHIP_NO_ERROR;
+}
+
+size_t GetPlusSignIdx(const ByteSpan & value)
+{
+    // Fist value is the vendor id, second (after the +) is the product.
+    for (int i = 0; i < static_cast<int>(value.size()); ++i)
+    {
+        if (static_cast<char>(value.data()[i]) == '+')
+        {
+            return i;
+        }
+    }
+    return value.size();
+}
+
+} // namespace
+
+uint16_t GetProduct(const ByteSpan & value)
+{
+    size_t plussign = GetPlusSignIdx(value);
+    if (plussign < value.size() - 1)
+    {
+        return MakeU16FromAsciiDecimal(ByteSpan(value.data() + plussign + 1, value.size() - plussign - 1));
+    }
+    return 0;
+}
+
+uint16_t GetVendor(const ByteSpan & value)
+{
+    size_t plussign = GetPlusSignIdx(value);
+    return MakeU16FromAsciiDecimal(ByteSpan(value.data(), plussign));
+}
+
+uint16_t GetLongDisriminator(const ByteSpan & value)
+{
+    return MakeU16FromAsciiDecimal(value);
+}
+
+uint8_t GetAdditionalPairing(const ByteSpan & value)
+{
+    return MakeU8FromAsciiDecimal(value);
+}
+
+uint8_t GetCommissioningMode(const ByteSpan & value)
+{
+    return MakeU8FromAsciiDecimal(value);
+}
+
+// TODO: possibly 32-bit? see spec issue #3226
+uint16_t GetDeviceType(const ByteSpan & value)
+{
+    return MakeU16FromAsciiDecimal(value);
+}
+
+void GetDeviceName(const ByteSpan & value, char * name)
+{
+    size_t len = std::min(value.size(), kMaxDeviceNameLen);
+    memcpy(name, value.data(), len);
+    name[len] = '\0';
+}
+
+void GetRotatingDeviceId(const ByteSpan & value, uint8_t * rotatingId, size_t * len)
```

Comment:
This would be best done as an analogue of src/lib/support/BytesToHex.h (e.g. HexToBytes()), since hex string ByteSpan to bytes buffer is used in several other parts of the spec.